server/jobs_controller.go

```diff
@@ -0,0 +1,201 @@
+// it, without telling any clients about it. This won't have any
+// effect on the frontend, except that newly connected clients won't
+// get the old messages sent to them, because the server has
+// "forgotten" about them. We clear the buffer for a PR when starting
+// and finishing processing that PR, because after the processing has
+// finished the output is posted to GitHub anyway and it doesn't need
+// to be persisted in the Atlantis web UI.
+
+type JobsController struct {
+	AtlantisVersion     string
+	AtlantisURL         *url.URL
+	Logger              *logging.SimpleLogger
+	JobTemplate         TemplateWriter
+	TerraformOutputChan <-chan *models.TerraformOutputLine
+
+	logBuffers map[string][]string
+	wsChans    map[string]map[chan string]bool
+	chanLock   sync.Mutex
+}
+
+// Register client to get Terraform output for PR. Send all the output
+// currently in the buffer.
+func (j *JobsController) addChan(pull string) chan string {
+	ch := make(chan string, 1000)
+	j.chanLock.Lock()
+	for _, line := range j.logBuffers[pull] {
+		ch <- line
+	}
+	if j.wsChans == nil {
+		j.wsChans = map[string]map[chan string]bool{}
+	}
+	if j.wsChans[pull] == nil {
+		j.wsChans[pull] = map[chan string]bool{}
+	}
+	j.wsChans[pull][ch] = true
+	j.chanLock.Unlock()
+	return ch
+}
+
+// Deregister client to no longer get Terraform output for PR.
+func (j *JobsController) removeChan(pull string, ch chan string) {
+	j.chanLock.Lock()
+	delete(j.wsChans[pull], ch)
+	j.chanLock.Unlock()
+}
+
+// Clear out log lines that are in the buffer (which would be
+// immediately received by a newly connecting client). This won't
+// cause anything to be deleted from an existing client webpage, but
+// it does mean the data won't reappear on a refresh.
+func (j *JobsController) clearLogLines(pull string) {
+	j.chanLock.Lock()
+	delete(j.logBuffers, pull)
+	j.chanLock.Unlock()
+}
+
+// Add a log line to the buffer and send it to all clients.
+func (j *JobsController) writeLogLine(pull string, line string) {
+	j.chanLock.Lock()
+	if j.logBuffers == nil {
+		j.logBuffers = map[string][]string{}
+	}
+	for ch, _ := range j.wsChans[pull] {
+		select {
+		case ch <- line:
+		default:
+			// If we get to this case then it means that
+			// the channel is full, which probably means
+			// that the client has disconnected, which
+			// results in items not getting pulled off the
+			// channel to send to the websocket, which
+			// means we should stop stuffing things into
+			// that channel.
+			delete(j.wsChans[pull], ch)
+		}
+	}
+	if j.logBuffers[pull] == nil {
+		j.logBuffers[pull] = []string{}
+	}
+	j.logBuffers[pull] = append(j.logBuffers[pull], line)
+	j.chanLock.Unlock()
+}
+
+// Listens to the Terraform output channel and distributes messages to
+// all clients. This function blocks, please invoke it in a goroutine.
+func (j *JobsController) Listen() {
+	for msg := range j.TerraformOutputChan {
+		if msg.ClearBefore {
```

Comment:
is this mainly invoked in the case that new plans are triggered for the same PR before a buffer is cleared?

Reply:
I don't think it should be necessary actually given the current implementation. I guess maybe in the case that a plan crashes and burns badly and the loop over the projects doesn't finish somehow. But I always like to clear caches at startup, not only teardown, "just in case".